File: code/A_Gentle_Introduction_to_Spark-Chapter_3_A_Tour_of_Sparks_Toolset.scala
// in Scala
import spark.implicits._
case class Flight(DEST_COUNTRY_NAME: String,
                  ORIGIN_COUNTRY_NAME: String,
                  count: BigInt)
val flightsDF = spark.read
  .parquet("/data/flight-data/parquet/2010-summary.parquet/")
val flights = flightsDF.as[Flight]


// COMMAND ----------

// in Scala
flights
  .filter(flight_row => flight_row.ORIGIN_COUNTRY_NAME != "Canada")
  .map(flight_row => flight_row)
  .take(5)

flights
  .take(5)
  .filter(flight_row => flight_row.ORIGIN_COUNTRY_NAME != "Canada")
  .map(fr => Flight(fr.DEST_COUNTRY_NAME, fr.ORIGIN_COUNTRY_NAME, fr.count + 5))


// COMMAND ----------

// in Scala
val staticDataFrame = spark.read.format("csv")
  .option("header", "true")
  .option("inferSchema", "true")
  .load("/data/retail-data/by-day/*.csv")

staticDataFrame.createOrReplaceTempView("retail_data")
val staticSchema = staticDataFrame.schema


// COMMAND ----------

// in Scala
import org.apache.spark.sql.functions.{window, column, desc, col}
staticDataFrame
  .selectExpr(
    "CustomerId",
    "(UnitPrice * Quantity) as total_cost",
    "InvoiceDate")
  .groupBy(
    col("CustomerId"), window(col("InvoiceDate"), "1 day"))
  .sum("total_cost")
  .show(5)


// COMMAND ----------

spark.conf.set("spark.sql.shuffle.partitions", "5")


// COMMAND ----------

val streamingDataFrame = spark.readStream
    .schema(staticSchema)
    .option("maxFilesPerTrigger", 1)
    .format("csv")
    .option("header", "true")
    .load("/data/retail-data/by-day/*.csv")


// COMMAND ----------

streamingDataFrame.isStreaming // returns true


// COMMAND ----------

// in Scala
val purchaseByCustomerPerHour = streamingDataFrame
  .selectExpr(
    "CustomerId",
    "(UnitPrice * Quantity) as total_cost",
    "InvoiceDate")
  .groupBy(
    $"CustomerId", window($"InvoiceDate", "1 day"))
  .sum("total_cost")


// COMMAND ----------

// in Scala
purchaseByCustomerPerHour.writeStream
    .format("memory") // memory = store in-memory table
    .queryName("customer_purchases") // the name of the in-memory table
    .outputMode("complete") // complete = all the counts should be in the table
    .start()


// COMMAND ----------

// in Scala
spark.sql("""
  SELECT *
  FROM customer_purchases
  ORDER BY `sum(total_cost)` DESC
  """)
  .show(5)


// COMMAND ----------

staticDataFrame.printSchema()


// COMMAND ----------

// in Scala
import org.apache.spark.sql.functions.date_format
val preppedDataFrame = staticDataFrame
  .na.fill(0)
  .withColumn("day_of_week", date_format($"InvoiceDate", "EEEE"))
  .coalesce(5)


// COMMAND ----------

// in Scala
val trainDataFrame = preppedDataFrame
  .where("InvoiceDate < '2011-07-01'")
val testDataFrame = preppedDataFrame
  .where("InvoiceDate >= '2011-07-01'")


// COMMAND ----------

trainDataFrame.count()
testDataFrame.count()


// COMMAND ----------

// in Scala
import org.apache.spark.ml.feature.StringIndexer
val indexer = new StringIndexer()
  .setInputCol("day_of_week")
  .setOutputCol("day_of_week_index")


// COMMAND ----------

// in Scala
import org.apache.spark.ml.feature.OneHotEncoder
val encoder = new OneHotEncoder()
  .setInputCol("day_of_week_index")
  .setOutputCol("day_of_week_encoded")


// COMMAND ----------

// in Scala
import org.apache.spark.ml.feature.VectorAssembler

val vectorAssembler = new VectorAssembler()
  .setInputCols(Array("UnitPrice", "Quantity", "day_of_week_encoded"))
  .setOutputCol("features")


// COMMAND ----------

// in Scala
import org.apache.spark.ml.Pipeline

val transformationPipeline = new Pipeline()
  .setStages(Array(indexer, encoder, vectorAssembler))


// COMMAND ----------

// in Scala
val fittedPipeline = transformationPipeline.fit(trainDataFrame)


// COMMAND ----------

// in Scala
val transformedTraining = fittedPipeline.transform(trainDataFrame)


// COMMAND ----------

transformedTraining.cache()


// COMMAND ----------

// in Scala
import org.apache.spark.ml.clustering.KMeans
val kmeans = new KMeans()
  .setK(20)
  .setSeed(1L)


// COMMAND ----------

// in Scala
val kmModel = kmeans.fit(transformedTraining)


// COMMAND ----------

kmModel.computeCost(transformedTraining)


// COMMAND ----------

// in Scala
val transformedTest = fittedPipeline.transform(testDataFrame)


// COMMAND ----------

kmModel.computeCost(transformedTest)


// COMMAND ----------

// in Scala
spark.sparkContext.parallelize(Seq(1, 2, 3)).toDF()


// COMMAND ----------

File: code/A_Gentle_Introduction_to_Spark-Chapter_3_A_Tour_of_Sparks_Toolset.py
# %%
from pyspark.sql import SparkSession
spark = SparkSession.builder.appName("spark-book").getOrCreate()

# %%
spark

# %%
staticDataFrame = (
    spark.read.format("csv")
    .option("header", "true")
    .option("inferSchema", "true")
    .load("../data/retail-data/by-day/*.csv")
)

staticDataFrame.createOrReplaceTempView("retail_data")
staticSchema = staticDataFrame.schema

# %%
from pyspark.sql.functions import window, column, desc, col

(staticDataFrame
    .selectExpr(
    "CustomerId", "(UnitPrice * Quantity) as total_cost", "InvoiceDate")
    .groupBy(col("CustomerId"), window(col("InvoiceDate"), "1 day"))
    .sum("total_cost")
    .show(5)
)

# %%
streamingDataFrame = (
    spark.readStream
    .schema(staticSchema)
    .option("maxFilesPerTrigger", 1)
    .format("csv")
    .option("header", "true")\
    .load("/data/retail-data/by-day/*.csv")
)

# %%
purchaseByCustomerPerHour = (
    streamingDataFrame
    .selectExpr("CustomerId", "(UnitPrice * Quantity) as total_cost", "InvoiceDate")
    .groupBy(col("CustomerId"), window(col("InvoiceDate"), "1 day"))
    .sum("total_cost")
)

# %%
purchaseByCustomerPerHour.writeStream\
    .format("memory")\
    .queryName("customer_purchases")\
    .outputMode("complete")\
    .start()

# %%
spark.sql("""
  SELECT *
  FROM customer_purchases
  ORDER BY `sum(total_cost)` DESC
  """)\
  .show(5)

# %%
from pyspark.sql.functions import date_format, col
preppedDataFrame = (
    staticDataFrame
    .na.fill(0)
    .withColumn("day_of_week", date_format(col("InvoiceDate"), "EEEE"))
    .coalesce(5)
)

# %%
trainDataFrame = (
    preppedDataFrame
    .where("InvoiceDate < '2011-07-01'")
)

testDataFrame = (
    preppedDataFrame
    .where("InvoiceDate >= '2011-07-01'")
)

# %%
from pyspark.ml.feature import StringIndexer
indexer = (
    StringIndexer()
    .setInputCol("day_of_week")
    .setOutputCol("day_of_week_index")
)

# %%

from pyspark.ml.feature import OneHotEncoder
encoder = (
    OneHotEncoder()
    .setInputCol("day_of_week_index")
    .setOutputCol("day_of_week_encoded")
)

# %%
from pyspark.ml.feature import VectorAssembler

vectorAssembler = (
    VectorAssembler()
    .setInputCols(["UnitPrice", "Quantity", "day_of_week_encoded"])
    .setOutputCol("features")
)

# %%
from pyspark.ml import Pipeline

transformationPipeline = (
    Pipeline()
    .setStages([indexer, encoder, vectorAssembler])
)

# %%
fittedPipeline = transformationPipeline.fit(trainDataFrame)

# %%
transformedTraining = fittedPipeline.transform(trainDataFrame)

# %%
from pyspark.ml.clustering import KMeans
kmeans = KMeans()\
  .setK(20)\
  .setSeed(1)

# %%
kmModel = kmeans.fit(transformedTraining)

# %%
transformedTest = fittedPipeline.transform(testDataFrame)

# %%
from pyspark.sql import Row

spark.sparkContext.parallelize([Row(1), Row(2), Row(3)]).toDF()

# %%
